sqlitecluster/SQLite.h

```diff
@@ -27,25 +26,31 @@ class SQLite {
         const char* what() const noexcept { return "checkpoint_required"; }
     };
 
+    // Abstract base class for objects that need to be notified when we set `checkpointRequired` and then when that
+    // checkpoint is complete.
+    class CheckpointRequiredListener {
+      public:
+        virtual void checkpointRequired(SQLite& db) = 0;
+        virtual void checkpointComplete(SQLite& db) = 0;
+    };
+
     // This publicly exposes our core mutex, allowing other classes to perform extra operations around commits and
     // such, when they determine that those operations must be made atomically with operations happening in SQLite.
     // This can be locked with the SQLITE_COMMIT_AUTOLOCK macro, as well.
-    static SLockTimer<recursive_mutex> g_commitLock;
+    static recursive_mutex& g_commitLock;
 
-    // Loads a database and confirms its schema
-    // The journalTable and numJournalTables parameters are maybe less than straightforward, here's what they mean:
-    //
-    // journalTable: this is the numerical id of the journalTable that this DB will *write* to. It can be -1 to
-    //               indicate that the table to use is 'journal', otherwise it will be journalNNNN, where NNNN is the
-    //               integer passed in. The actual table name use will always have at least four digits (leading 0s
-    //               for numbers less than 1000).
-    //
-    // maxRequiredJournalTableID: This is the maximum journal table ID that we'll verify. If it's -1, we'll only verify
-    //                            'journal' and no numbered tables.
+    // createJournalTables: Creates the specified number of journal tables. If `0`, no tables are created. This
+    //                      specifies the total number of journal tables, not new ones, so if there are 50 existent
+    //                      tables and you pass `100` here, you get 100 total tables, not 150. 
     //
     // mmapSizeGB: address space to use for memory-mapped IO, in GB.
-    SQLite(const string& filename, int cacheSize, bool enableFullCheckpoints, int maxJournalSize, int journalTable,
-           int maxRequiredJournalTableID, const string& synchronous = "", int64_t mmapSizeGB = 0, bool pageLoggingEnabled = false);
+    SQLite(const string& filename, int cacheSize, bool enableFullCheckpoints, int maxJournalSize,
+           int createJournalTables, const string& synchronous = "", int64_t mmapSizeGB = 0, bool pageLoggingEnabled = false);
+
+    // This constructor is not exactly a copy constructor. It creates an other SQLite object based on the first except
+    // with a *different* journal table. This avoids a lot of locking around creating structures that we know already
+    // exist because we already have a SQLite object for this file.
+    SQLite(const SQLite& from);
```

Comment:
This is used to create all of the handles to a DB except the first one. Creating the first one does a whole bunch of extra work that's all guarded inside various locks to make sure it only happens once. For instance, in the other SQLite constructor, we do:

```
    // We lock here to initialize the database. Because there's a global map of currently opened DB files, we lock
    // whenever we might need to insert a new one. These are only ever added or changed in the constructor and
    // destructor.
    SQLITE_COMMIT_AUTOLOCK;

more work here...
```

So, while creating a SQLite object, we block commits. We only need to do this for the first handle to a DB, but we have up to now done it for all of them, because we only ever make these handles at startup.

Now that we create handles arbitrarily any time, it'd be nice to create DB handles while skipping the expensive locking and such that only needs to happen for the first one. By creating a DB handle based on an existing handle to the same DB, we can skip all the "first one" logic, as we know there's already at least one handle to this DB (the one we're initializing based upon).